napari/layers/image/image.py

```diff
@@ -475,51 +475,43 @@ def _set_view_slice(self):
 
-            disp_shape = self.level_shapes[level, self.dims.displayed]
             scale = np.ones(self.ndim)
             for d in self.dims.displayed:
                 scale[d] = self.level_downsamples[self.data_level][d]
             self._transforms['tile2data'].scale = scale
 
-            if np.any(disp_shape > self._max_tile_shape):
-                for d in self.dims.displayed:
-                    indices[d] = slice(
-                        self._top_left[d],
-                        self._top_left[d] + self._max_tile_shape,
-                        1,
-                    )
-                # Note that top left marks the location of top left canvas
-                # pixel in data coordinates
-                self._transforms['tile2data'].translate = (
-                    self._top_left
-                    * self._transforms['data2world'].scale
-                    * self._transforms['tile2data'].scale
+            for d in self.dims.displayed:
+                indices[d] = slice(
+                    self.corner_pixels[0, d], self.corner_pixels[1, d] + 1, 1,
                 )
-            else:
-                self._transforms['tile2data'].translate = np.zeros(self.ndim)
+            self._transforms['tile2data'].translate = (
+                self.corner_pixels[0]
+                * self._transforms['data2world'].scale
+                * self._transforms['tile2data'].scale
+            )
 
             image = np.asarray(
                 self._data_pyramid[level][tuple(indices)]
             ).transpose(order)
 
-            if level == len(self._data_pyramid) - 1:
-                thumbnail = image
-            else:
-                # Slice thumbnail
-                indices = np.array(self.dims.indices)
-                downsampled_indices = (
-                    indices[not_disp] / self.level_downsamples[-1, not_disp]
-                )
-                downsampled_indices = np.round(
-                    downsampled_indices.astype(float)
-                ).astype(int)
-                downsampled_indices = np.clip(
-                    downsampled_indices, 0, self.level_shapes[-1, not_disp] - 1
-                )
-                indices[not_disp] = downsampled_indices
-                thumbnail = np.asarray(
-                    self._data_pyramid[-1][tuple(indices)]
-                ).transpose(order)
+            if np.product(image.shape) == 0:
+                image = np.zeros((1, 1))
+
+            # Slice thumbnail
+            indices = np.array(self.dims.indices)
+            downsampled_indices = (
+                indices[not_disp] / self.level_downsamples[-1, not_disp]
+            )
+            downsampled_indices = np.round(
+                downsampled_indices.astype(float)
+            ).astype(int)
+            downsampled_indices = np.clip(
+                downsampled_indices, 0, self.level_shapes[-1, not_disp] - 1
+            )
+            indices[not_disp] = downsampled_indices
```

Comment:
This is a heck of a lot of work for a thumbnail, and I expect that this work is also done elsewhere? Could we not reuse that code/computation?

I also generally dislike reusing a variable name over and over during processing, and, given black formatting, I might prefer a single call:

```python
# find index in the current pyramid level
downsampled_indices = np.clip(
    np.round(
        indices[not_disp] / self.downsample_factors[-1, not_disp]
    ).astype(int),
    0,
    self.level_shapes[-1, not_disp] - 1,
)
```

Reply:
Regarding reuse, I'm thinking something like, `def slice_index_convert_to_level`, then use that function with `level=-1`. I presume you would use it when actually rendering on the canvas? **Edit:** Yeah, I see that you do the exact same thing above but with the current level rather than the smallest level.

Reply:
can i fix this in the next PR which will remove pyramid autogeneration and will make this stuff much easier to clean up as won't have this weird case where we were given an array but are using a pyramid. will be much nicer then